### ISSUE TITLE
@Secured annotations, recent changes and question

### ISSUE THREAD

**Issue by FiXerCz** [Cweorth/Surreal-Travel#47]
Issue regarding my commit ddc3b4ea5fc0f4aff613b95122658745ba3a9d72. Second opinion wanted :)

*This commit changed quite a few things:*

1) In SpringSec I have enabled annotations @Secured and @RolesAllowed. Those can be used throughout the code to restrict user roles, that should have access to the annotated methods.

2) I have added error messages when user inputs wrong credentials (forgot about it before :smile: ). Probably nothing can go wrong with this change.

3) ```AuthController.java``` is also pretty self-explanatory and easy enough in order for me not to screw it up.

4) ```AuthCommons.java``` - there would be quite lot of repeating of SpringSec code calls just to get a simple username or determining which roles (authorities) user has. This could simplify it for us.

5) ReservationController + its views have quite a few changes and that is the part, where I would like your opinion (when you have the time).

---

**ReservationController:**

I have annotated all methods according to roles that should have access to them. Basically the whole reservation-part-of-app should be only accessible for authenticated people (unlike Excursions and Trips - listing could be shown to all), so I have hidden it from the menu for anonymous users. Some methods should have different functionality for ROLE_USER and for higher roles. I have dealt with this as best as I knew how.

Basically for ROLE_STAFF and ROLE_ADMIN ReservationController is the same as it was before commit. For ROLE_USER:
- Only user's reservations are shown
- User can create reservation only for himself (no Customer select box)
- Should also ROLE_USER be able to edit and delete his reservations (right now the funcionaly is hidden from GUI and reserved for STAFF and ADMIN only)?

Do you think this approach will be fine or do you see any holes in the current version?

Oh and a footnote: Login page is no longer at /login but at /auth/login. Could you please change your redirect url after creating new Account? Thanks and sorry about it :smile:

**Comment by Cweorth** [comment]
Ad ```@Secured```, shouldn't that be on Service layer? Or are there other annotations that should be used there?
Ad ```AuthCommons.java```, it is a good idea.
Ad ```ReservationController```, these changes are good. 

I think it does not really matter if an user can delete his own reservations, both *yes* or *no* can be justified... but this depends on other factors too, like "**Do we permit unregistration**"? 
Right now, ```AccountController.deleteAccount()``` method (only in my local repo) will remove all reservations, the customer (if not used elsewhere) and the account (unless he's the last admin). Admins will definitely have access to that functionality for all users, but if we allow an user to unregister himself, it logically implies we need to allow cancellation (other direction does not work though).

In real life, cancelling a reservation is usually possible for a explicitly stated period of time (like "until 3 days before departure (inclusive)"). Since none of our trips will (or **could** for that matter) actually take place (though it makes me sad to think I will never see Rivendell), I think we should enable this.

**Comment by FiXerCz** [comment]
> Ad @Secured, shouldn't that be on Service layer? Or are there other annotations that should be used there?

If we put those annotations on the service layer, then user that has no access to for example editing Reservations (reservationService.editReservation(...)) could still access page with form for editing reservations in web. We would either have to keep annotations both in service and web or dunno..

Also, when I tried just now to move annotations to Service tier, I am getting several exceptions about non-existent authentication object. Probably because authentication took place in web and then I try to utilize it in Service. Maybe it could be achieved, but we would need to do some more research on google.

And lastly, if we put annotations to Service tier, how would we then proceed with with Rest client, that does not need to have authentication? Would those methods still be accessible?

> Since none of our trips will (or could for that matter) actually take place (though it makes me sad to think I will never see Rivendell), I think we should enable this.

Alrighty, that makes sense. I will enable it then. After all, it is not that difficult to do it.

**Comment by Cweorth** [comment]
Of course I agree; your point does make sense. But then the following sentence

> Implement authentication and authorization on both presentation and **service layer** of
the web application

does not. And as someone pointed out on FB discussion group (which does not make it right, but can be a basis for a research)

> Anotacie na service beanach pre uzivatelske role.

which can be confusing, since it is a mix of Slovak, Czech and English (either way, *wrong*), so I shall translate that:
> *(sk)* Anotacie na servisnych beanoch pre uzivatelske roly.
> *(cz)* Anotace na servisnich beanech pro uzivatelske role.
> *(en)* Annotations on service beans for user roles.
> *(de)* Anmerkungen auf Service Beanen für Benutzerrollen.

My point is, we are expected to implement authentication on service layer **somehow**, and I thought that on Web, we should *hide* the privileged operations and on service layer we should *check* them.
Like, even if you hide a link to ```trips/delete/5```, you can still type it out, and there is where service layer should stop you. Or I don't know :goberserk:

----

**BTW**, the requirements for REST are quite contradictory.

> REST client does not need to be authenticated, to be able to access **protected service
layer**, default **username “rest” with password “rest” should be hardcoded** in the
implementation of the REST/SOAP API layer.

So, if the client is not authenticated, why do you need the username and password for it? I think they actually meant that the **user** does not need to provide the credentials to use the application, as the application will provide it on its own.

Also, notice that it mentions a **protected service layer** again.
It's kind of mysterious I think.

**Comment by Cweorth** [comment]
I've just realized another thing: 

> Implement [*that crap*] on both [...] layer[*s*] **of the web application**

this could also mean that we are not expected to directly modify the service layer, but instead, create a proxy service objects (using AOP) where we would be able to access security context.

**Comment by FiXerCz** [comment]
Yep, I also am lost. If the tutor could shed some light onto this it would help..

**Comment by Cweorth** [comment]
Ok, I will write him. 

(Oh god, the 4th paper from IA006 has been scanned into the IS :scream: ... I butchered that nicely)

**Comment by Cweorth** [comment]
Response from our tutor:

> autorizácia by mala prebiehať ako na úrovni prezentačnej vrstvy, 
tak aj na úrovni business logiky (a je to všeobecne dobrá stratégia).

> Prezentačná vrstva - overenie, či sa užívateľovi môže zobraziť daná 
stránka. Servisná vrstva - overenie, či užívateľ môže skutočne vykonať 
danú operáciu (napr. aj keď pristupuje priamo cez nejakého klienta).

So, what now? :smiley:

**Comment by FiXerCz** [comment]
Well that was very helpfull, wasn't it? :D

I suppose we need to consult with uncle Google. I think I have already seen such approach somewhere. Though maybe it was "move authentization to service" and not "lets have authentization everywhere"...

BTW I have messed the 4th paper as well - I still need 19 out of 30 possible points from the first one.. Damn...

**Comment by Cweorth** [comment]
Mmm, Security context is a static class, right? I think it should be possible to add Spring Security dependency to the project and then either rewrite the service layer or use AOP (to avoid breaking what already works :smiley:)

(I'm "only" 4 points short, but since I failed to show that the language is LALR(1) and then I succeeded in creating SLR(1) automaton (which should not be possible), it would be a miracle to get these points :disappointed: )

**Comment by FiXerCz** [comment]
I have messed around with securing of Service tier just about now and turns out, it worked just fine before. I hadn't notice, that the exception had complained about calling secured service method ```addReservation(Reservation)``` in ```@PostConstruct``` initialization method and that call is, of course, not authenticated. I will try to find a way around it. I have also found commented out code in DefaultReservationService - we were quite lucky no one noticed that, we could have lost some points there. I have removed it and will fix it in next commit.

But, generally speaking, securing service tier will not be a problem.

PS: So how did you do with automatons? I had more luck than sense - my whole semestral points sum up to exactly 55 points. Hell of a luck... :)

**Comment by Cweorth** [comment]
Ok, when you find a working solution, push it, so we can bring this folly to an end.

> I have also found commented out code in DefaultReservationService

You mean line ```79```? I probably missed it when checked the code last time. But it's only one line that... I have no idea what it was supposed to do :smiley:

Ad automatons: I got 19 pts from the first paper, but only because the grader **may** have missed an error there (but if anyone asks, she did not :speak_no_evil: ) so overall, I got 71 @ **D**.

**Comment by FiXerCz** [comment]
I already have a solution. It was a bitch. It is neccessary to do "manual" authentication at the beggining of the PostContruct method in InitDataController. In order to do authentication, one needs some user in DB already, so I had hibernate on DAO layer run simple sql script to store "pa165" user and that user is then used for saving all our initial Customers, Excursions etc.

I am currently moving all init methods from controllers to InitDataController. I will do Customer, Excursion, Reservation. Account is already there and you can do Trip, so that I dont modify your code.

I will push it asap.

---

Yeah, line 79. I also cannot fathom its meaning. :)

---

That is great! Important thing is that it is green and not red in the IS! :) I really did not feel like going to another exam..

**Comment by Cweorth** [comment]
Ad ```init```, this solution will serve for now. But later, when we will finalize the code, we should make initial credentials to ```root``` or ```administrator``` (```pa165``` is not a username one would expect to be the master admin :smiley: ) and add ```pa165``` after that.

----

Ok, before pushing, please fix ```auth.wrongCredentials``` key (*Invalid username __of__ password*).

**Comment by Cweorth** [comment]
Fuck, I forgot I had changed a line in ```AuthCommons``` and I pushed it :disappointed: it's just a check (```SecurityContextHolder.getContext().getAuthentication()``` can return ```null```), but if it collides, remove it. Sorry

**Comment by FiXerCz** [comment]
No problem, I do not have modifications there.

**Comment by FiXerCz** [comment]
Ok, big commit is live. I have changed several things so I tried to double check. Hopefully i did not miss anything.

We can "normalize" content of InitDataController in final check ups l8r. I did not care for content much now (eg Biking in Haradwaith offers excursions to Afganistan...).

**Comment by Cweorth** [comment]
lol, that's quite redundant, because if Afganistan were in Middle-Earth, it would be called Haradwaith :smile:

**Comment by FiXerCz** [comment]
:D

**Comment by FiXerCz** [comment]
Just a note about just closed topic, I am currently finishing Excursion and Customer parts (both presentation and service) so dont do those, it will be live shortly.

I did not know I should wait so i have already started :D

**Comment by Cweorth** [comment]
Oh, dammit, I thought you'd do Reservation instead of Excursion. Ok, I will do it instead :D

**Comment by FiXerCz** [comment]
Well Reservation is also almost complete :D I can do all 3 (Customer, Excursion, Reservation). Account is quite complicated so I thought it would be fair if I do those 3 and leave you Account and your Trip.

**Comment by Cweorth** [comment]
Well, I can't forbid you to do more work, especially willingly, so ... all right :+1: :smiley:
